contracts/axelar-gateway/src/auth.rs

```diff
@@ -235,3 +235,656 @@ fn validate_signers(env: &Env, weighted_signers: &WeightedSigners) -> Result<(),
+
+        // validate_proof should panic, signatures do not match signers
+        env.as_contract(&contract_id, || {
+            assert_err!(
+                auth::validate_proof(&env, &msg_hash, proof),
+                ContractError::InvalidSignersHash
+            );
+        })
+    }
+
+    #[test]
+    fn fail_validate_proof_threshold_not_met() {
+        let (env, contract_id, client) = setup_env();
+        let owner = Address::generate(&env);
+        let operator = Address::generate(&env);
+
+        let signers = initialize(
+            &env,
+            &client,
+            owner,
+            operator,
+            randint(0, 10),
+            randint(1, 10),
+        );
+
+        let mut total_weight = 0u128;
+
+        let msg_hash: BytesN<32> = BytesN::random(&env);
+        let mut proof = generate_proof(&env, msg_hash.clone(), signers);
+
+        // Modify signatures to make them invalid
+        let mut new_signers = Vec::new(&env);
+        for ProofSigner { signer, signature } in proof.signers.iter() {
+            total_weight += signer.weight;
+
+            if total_weight < proof.threshold {
+                new_signers.push_back(ProofSigner { signer, signature });
+            } else {
+                new_signers.push_back(ProofSigner {
+                    signer,
+                    signature: ProofSignature::Unsigned,
+                });
+            }
+        }
+        proof.signers = new_signers;
+
+        // should panic, all signatures are valid but total weight is below threshold
+        env.as_contract(&contract_id, || {
+            assert_err!(
+                auth::validate_proof(&env, &msg_hash, proof),
+                ContractError::InvalidSignatures
+            );
+        })
+    }
+    #[test]
+    fn fail_validate_proof_threshold_overflow() {
+        let (env, contract_id, client) = setup_env();
+        let owner = Address::generate(&env);
+        let operator = Address::generate(&env);
+
+        let mut signers = initialize(
+            &env,
+            &client,
+            owner,
+            operator,
+            randint(0, 10),
+            randint(1, 10),
+        );
+
+        let last_index = signers.signers.signers.len() - 1;
+
+        // get last signer and modify its weight to max u128 - 1
+        if let Some(mut last_signer) = signers.signers.signers.get(last_index) {
+            last_signer.weight = u128::MAX - 1;
+            signers.signers.signers.set(last_index, last_signer);
+        }
+
+        let msg_hash: BytesN<32> = BytesN::random(&env);
+        let proof = generate_proof(&env, msg_hash.clone(), signers);
+
+        // should panic, as modified signer wouldn't match the epoch
+        env.as_contract(&contract_id, || {
+            assert_err!(
+                auth::validate_proof(&env, &msg_hash, proof),
+                ContractError::InvalidSignersHash
+            );
+        });
+    }
+
+    #[test]
+    fn test_rotate_signers() {
```

Comment:
avoid `test_` prefix
```suggestion
    fn rotate_signers_succeeds() {
```

Reply:
done in [70f1b26](https://github.com/axelarnetwork/axelar-cgp-soroban/pull/49/commits/70f1b26bdb431ad1e877243c5aa318f0a9d5cf5f)